### DISCUSSION TITLE
Refreshing almost never happens automatically

### DISCUSSION THREAD

**Discussion by Gabriel-p** [elementary/files#2355] [General]
### What Happened?

Lately I've found that Files has become almost unusable unless I am constantly manually refreshing after any change made. File deletion, creation, renaming, anything. Unless I manually refresh, the changes are not shown properly in Files. With images I usually have to manually refresh up to ten times before the thumbnails are shown.

This is just a minor example. This PDF file was just created by compiling with Latex. As seen, the icon is not that of a PDF file. This is a minor issue if not for the fact that attempting to open that file loads up the text editor, not the PDF viewer.

![Captura de pantalla de 2023-08-30 17 20 00](https://github.com/elementary/files/assets/6054951/f68bfc13-3b15-4316-9681-7c38e37fef9b)

Only after refreshing does the proper thumbnail show and I can open the file with a double click

![Captura de pantalla de 2023-08-30 17 21 49](https://github.com/elementary/files/assets/6054951/a187c8aa-5ed3-4adc-bbcf-0e10b0cc4777)

### Steps to Reproduce

Use Files 

### Expected Behavior

This should not happen

### OS Version

7.x (Horus)

### Software Version

Latest release (I have run all updates)

**Comment by jeremypw** [comment]
Version 6.5.0 contains a fix for a regression in refreshing the view when a file is added.  Which version are you running?

**Comment by jeremypw** [comment]
The next version will contain a fix for regressions after a file is renamed.

**Comment by Gabriel-p** [comment]
> Version 6.5.0 contains a fix for a regression in refreshing the view when a file is added. Which version are you running?

I'd like to tell you but apparently there's no way to see the version in the menu (why?) and I don't know the name of the app to use a terminal command. Applications says I'm using version 6.4.1, not sure if that's the right one

**Comment by jeremypw** [comment]
You should be able to get the version with the command `io.elementary.files --version` but the AppCenter should be accurate provided you have rebooted or logged out since updating to make sure the running processes are updated.  We do not put things in the app interface that are not directly needed for controlling the app.

**Comment by Gabriel-p** [comment]
The refreshing issue is still there for images. These are png files created by a Python script, before refreshing they are identified as text files:

![Captura de pantalla de 2023-09-05 17 49 16](https://github.com/elementary/files/assets/6054951/9daf3adc-0e64-4c09-a7f3-a246a524f522)

Only after manually refreshing are they recognized as images:

![Captura de pantalla de 2023-09-05 17 49 28](https://github.com/elementary/files/assets/6054951/37e66969-aa75-459a-8fc8-9f22c617330f)

Thumbnails do not appear unless I manually refresh multiple times. Running version 6.5.0

**Comment by jeremypw** [comment]
@Gabriel-p  Thanks for the update but thumbnailing is controlled by different code so this needs to be raised as a separate issue (if not already done).

**Comment by Gabriel-p** [comment]
It's not just the thumbnailing, that's a minor issue, the problem is that the files are not recognized as what they are (images in this case) until I manually refresh. Until I do, the system attempts to open them as text files. Unless this issue is part of the no thumbnailing issue...

**Comment by jeremypw** [comment]
It may be related - certain information (like thumbnails) is not loaded until the file is actually displayed in order to improve performance.  I'll open a new issue.

**Comment by zappan** [comment]
@jeremypw without rushing to open a new issue, as it seems to be related with the refresh:

Creating a new folder by using a keyboard shortcut, naming it (name change is properly reflected under the folder icon), and then entering it with the Enter key, ends up in the 'untitled folder' in the location bar (example: `/home/<username>/tmp/untitled folder`), and of course, pasting a previously copied or cut content fails due to that. Getting 'up' a folder via keyboard, and then selecting it and entering via Enter works correctly, ending up in a folder as it was named (example: `/home/<username>/tmp/thisfolder`). I'd assume a refresh happened when going in and out/up again.

If I were to hit a CTRL+R after renaming the folder, it would work properly.

Do let me know if I need to provide formal reproduction steps

```
$ io.elementary.files --version
io.elementary.files 6.5.0
```

**Comment by jeremypw** [comment]
@zappan I could not reproduce this using Files compiled from source. Could you provide exact steps and I'll retry with the released version.

The pathbar can only show the path that was *requested* which should be the path of the selected fileitem when `Enter` is pressed.  If that path gets renamed on disk *after* the request (due to some timing issue) then you should get the "This Folder Does Not Exist" warning.

**Comment by jeremypw** [comment]
Update:  I have confirmed the issue in released v 6.5.0 but it is fixed in `main` by d9cef820ea7f8af7100947aeb2575ce1fb039680

I'll get another release out soon as there is another important fix that needs releasing.

**Comment by peteruithoven** [comment]
@Gabriel-p any change you can share an easy way to reproduce these issues? Something we can easily try without using Latex, or a Python script that generates images for example. Or maybe share the script you use to create the images? 

With the current version of files I have a test.png file in a folder.
When I run `cp test.png test2.png` using the Terminal the new file appears as an image file. When I double click it, it opens in an image viewer. 
It only doesn't get a thumbnail unless I refresh. 
![Screenshot from 2023-10-01 19 15 28@2x](https://github.com/elementary/files/assets/523210/c9b08240-2057-468f-bb64-89fa640e1ce7)
Which makes me think @Gabriel-p's issue is solved.

With the latest version I can't reproduce @zappan's issue.

**Comment by Gabriel-p** [comment]
The issue is still there. Here's a way I can reproduce it. Generate a new .py file with the content:

```
import matplotlib.pyplot as plt
plt.scatter(0.5, 0.5)
plt.savefig('test.png')
```

Run the script to generate the `test.png` file. This is what I get:

![Captura de pantalla de 2023-10-03 09 19 59](https://github.com/elementary/files/assets/6054951/4b41e3b5-e2e6-44a7-929e-c2dcd60f9518)

Notice neither file has the proper thumbnail. If I attempt to open the .png file this is what I get:

![Captura de pantalla de 2023-10-03 09 20 07](https://github.com/elementary/files/assets/6054951/c696cc08-53a4-4723-bad0-2ec27c5cded3)

I.e., it tries to open it as a text file. After manually refreshing the folder, I see:

![Captura de pantalla de 2023-10-03 09 20 17](https://github.com/elementary/files/assets/6054951/131a324a-0abe-434a-b205-1ee327b289f2)

and now I can open the .png file properly.

**Comment by jeremypw** [comment]
@Gabriel-p  This looks like a duplicate of #2312?   I am inclined to close this issue because it encompasses several different (but related) issues that should be opened separately where not already fixed.

**Comment by Gabriel-p** [comment]
I think that issue initially was about the thumbnails not being loaded,  but now it does look like a duplicate of this one. My comment above might need to be copied to that thread

**Comment by zappan** [comment]
@jeremypw @peteruithoven 

```
$ io.elementary.files --version
io.elementary.files 6.5.2
```

Confirming the issue I had does not occurre any more. 
Thanks

**Comment by redgpu** [comment]
```
$ io.elementary.files --version
io.elementary.files 6.5.2
```
I do have this issue still, though.

**Comment by redgpu** [comment]
I have to refresh each time anything is changed: a folder deleted, or an archive extracted, etc.

**Comment by jeremypw** [comment]
@redgu I could not reproduce the issues with deleting folders or extracting archives using version 6.5.2.  However there are a couple of refreshing issues whose fixes are waiting to be merged and released - https://github.com/elementary/files/pull/2340 and https://github.com/elementary/files/pull/2265.  If you do still have an issues that has not already been raised please raise a new report for each specific one as they may not have the same cause. Please give details steps to reproduce.

**Comment by jeremypw** [comment]
Converted to discussion as it covers a number of related issues.

**Comment by redgpu** [comment]
> Please give details steps to reproduce.

I installed elementary os 7.1, I started deleting some folders and extracting some archives on disk and noticed the Files never refreshed on these changes, I started googling who else has this problem, and the https://github.com/elementary/files/issues/2304 issue popped up.

**Reply by jeremypw** [reply]
Files relies on GLib.FileMonitors to signal when directories that are being viewed are changed by external events (in particular when files are added or removed from them). This only works on local directories - are the directories that are not refreshing local? As far as I know Nautilus works the same way so I am not sure why you are getting this problem. Does the problem occur when adding/removing regular files as well as folders?  Does it occur in all view modes?

Hidden files will not appear when added if "Show hidden files" is OFF. "Hidden files" includes those have "~" as a prefix or suffix as well as those with "." as prefix.

**Comment by redgpu** [comment]
Meanwhile, I will switch Applications -> Default Apps -> File Browser from Files to Nautilus to make it work for now.

```
sudo apt install nautilus nautilus-extension-gnome-terminal
nautilus -q
sudo apt remove pantheon-files
```

**Comment by Ppamungkas** [comment]
The issue is also happen when creating a screen cast using OBS. The file created need to be refreshed before recognized as a video file.
![image](https://github.com/elementary/files/assets/36324190/9552ff79-6a18-43ee-b1b3-cfa2d1169dce)
![image](https://github.com/elementary/files/assets/36324190/dcdf141e-4f02-4efd-a1ca-dd07dbea30a5)

**Comment by jeremypw** [comment]
Please make sure specific focused issues are opened in relation to subjects discussed here so that they may be addressed individually and maybe added to the Performance project.
